spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/reactive/PubSubReactiveFactory.java

```diff
@@ -0,0 +1,153 @@
+import org.springframework.cloud.gcp.pubsub.support.AcknowledgeablePubsubMessage;
+
+/**
+ * Creates a GCP Pub/Sub Subscription-backed Flux respecting backpressure by way of Pub/Sub Synchronous Pull.
+ *
+ * @author Elena Felder
+ *
+ * @since 1.2
+ */
+public final class PubSubReactiveFactory {
+
+	private static final Log LOGGER = LogFactory.getLog(PubSubReactiveFactory.class);
+
+	private PubSubSubscriberOperations subscriberOperations;
+
+	/**
+	 * Instantiates `PubSubReactiveFactory` capable of generating subscription-based streams.
+	 *
+	 * @param subscriberOperations template for interacting with GCP Pub/Sub subscriber operations.
+	 */
+	public PubSubReactiveFactory(PubSubSubscriberOperations subscriberOperations) {
+		this.subscriberOperations = subscriberOperations;
+	}
+
+	/**
+	 * Creates an infinite stream {@link Publisher} of {@link AcknowledgeablePubsubMessage} objects.
+	 * <p>For unlimited demand, the underlying subscription will be polled at a regular interval.
+	 * <p>For specific demand, as many messages as are available will be returned immediately, with remaining demand being
+	 * fulfilled in the future. Pub/Sub timeout will cause a retry with the same demand.
+	 * @param subscriptionName subscription from which to retrieve messages.
+	 * @param pollingPeriod how frequently to poll the source subscription in case of unlimited demand.
+	 * @return infinite stream of {@link AcknowledgeablePubsubMessage} objects.
+	 */
+	public Publisher<AcknowledgeablePubsubMessage> createPolledPublisher(String subscriptionName, long pollingPeriod) {
+
+		return Flux.<AcknowledgeablePubsubMessage>create(sink -> {
+			sink.onRequest((numRequested) -> {
+				if (numRequested == Long.MAX_VALUE) {
+					// unlimited demand
+					Disposable task = Schedulers.single().schedulePeriodically(
+							new PubSubNonBlockingUnlimitedDemandPullTask(subscriptionName, sink), 0, pollingPeriod, TimeUnit.MILLISECONDS);
+					sink.onCancel(() -> {
+						task.dispose();
+					});
+				}
+				else {
+					Schedulers.single().schedule(new PubSubBlockingLimitedDemandPullTask(subscriptionName, numRequested, sink));
+				}
+			});
+
+		});
+	}
+
+	private abstract class PubSubPullTask implements Runnable {
+
+		protected final String subscriptionName;
+
+		protected final FluxSink<AcknowledgeablePubsubMessage> sink;
+
+		PubSubPullTask(String subscriptionName, FluxSink<AcknowledgeablePubsubMessage> sink) {
+			this.subscriptionName = subscriptionName;
+			this.sink = sink;
+		}
+
+		protected List<AcknowledgeablePubsubMessage> pullToSink(long demand, boolean block) {
+			int numMessagesToPull = demand > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) demand;
+
+			List<AcknowledgeablePubsubMessage> messages =  PubSubReactiveFactory.this.subscriberOperations.pull(
+					this.subscriptionName, numMessagesToPull, !block);
+
+			if (messages.size() > 0) {
+				messages.forEach(sink::next);
+			}
+
+			return messages;
```

Comment:
I think we may end up with the contract just really return an amount we have just pulled, not the whole list of messages.